### PR TITLE
Scaleway: Fix DatasourceScaleway to avoid backtrace

### PR DESCRIPTION
**Pull request by karibou**
Make sure that network_config is created when self._network_config
is unset

### CODE REVIEW

File: tests/unittests/test_datasource/test_scaleway.py [hunk 1]
```diff
@@ -403,3 +404,25 @@ def test_network_config_existing(self, m_get_cmdline, fallback_nic):
 
         netcfg = self.datasource.network_config
         self.assertEqual(netcfg, '0xdeadbeef')
+
+    @mock.patch('cloudinit.sources.DataSourceScaleway.net.find_fallback_nic')
+    @mock.patch('cloudinit.util.get_cmdline')
+    def test_network_config_unset(self, m_get_cmdline, fallback_nic):
+        """
+        _network_config will be set to sources.UNSET after the first run.
```
Comment (smoser) [inline tests/unittests/test_datasource/test_scaleway.py]: that comment doesn't seem right.
what do you mean 'first run'?  after `cloud-init local` or second boot for the instance?

Either way, I think we should have cached the response somewhere.

maybe collect-logs on a failed instance  and attach?

**Comment by OddBloke** [comment]
Hey @karibou, thanks for the proposed change! We'll need you to sign the CLA before we can merge this, you can do so by following the process described in https://cloudinit.readthedocs.io/en/latest/topics/hacking.html. (If you've signed the CLA on Launchpad in the past, then that doc also describes how to transfer that signature over to GitHub.)

**Comment by karibou** [comment]
Hello Scott, OddBloke,

I knew I should have donc the PR on launchpad where my CLA is signed (louis@ubuntu.com). I'll transfer it over here.

Scott, I'll get a collect-logs tomorrow on a failing instance but what I meant is after cloud-init init.
It is basically the same problem than here : https://bugs.launchpad.net/cloud-init/+bug/1823084

thanks for looking into this,

...Louis Bouchard

**Comment by karibou** [comment]
Here is the cloud-init.tar.gz

For a reason that escapes my understanding, I'm unable to push my branch to LP as ~louis/cloud-init so the migration script cannot find it (it comes up as ~louis/+git/cloud-init). Can someone suggest the git URL to use for pushing my branch ?

I haven't done this in a few years so unfortunately I forgot.

...Louis

**Comment by OddBloke** [comment]
@karibou Can you paste the error you're seeing, please?

**Comment by karibou** [comment]
Hello,

As it turns out, it worked when I tried to reproduce the problem to give you the error so I was able to complete the process to connect lp CLI signed LP account to my GH account.

You should now have a Pull Request for that.

...Louis

File: tests/unittests/test_datasource/test_scaleway.py (hunk 1 above)
Comment (karibou) [reply tests/unittests/test_datasource/test_scaleway.py]: Well, looks like GH doesn't want my logs tarball. I'll try to find another way to make it available.

**Comment by karibou** [comment]
@OddBloke just so you know, my LP problem worked itself out so I was able to sync my CLA with GH.

**Comment by raharper** [comment]
> Scott, I'll get a collect-logs tomorrow on a failing instance but what I meant is after cloud-init init.
> It is basically the same problem than here : https://bugs.launchpad.net/cloud-init/+bug/1823084

Hrm, in that case DatasourceAzure implements check_instance_id() which allows cloud-init to cache the cloud object to disk;  the bug occurred when on subsequent boots we load the pickled cloud-object and in some scenarios the value was sources.UNSET versus None which broke the assumption that the value was trustable if not None.

DatasourceScaleway does not implement check_instance_id() which means the assumption that self._network_config is either None or valid should hold.  I cannot determine how the value would ever be sources.UNSET?

Can you file a launchpad bug with the traceback  error and logs?

**Comment by smoser** [comment]
the stack trace in question is this:

    2019-12-20 09:57:05,749 - util.py[WARNING]: failed stage init
    2019-12-20 09:57:05,750 - util.py[DEBUG]: failed stage init
    Traceback (most recent call last):
      File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 653, in status_wrapper
        ret = functor(name, args)
      File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 362, in main_init
        init.apply_network_config(bring_up=bool(mode != sources.DSMODE_LOCAL))
      File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 701, in apply_network_config
        net.wait_for_physdevs(netcfg)
      File "/usr/lib/python3/dist-packages/cloudinit/net/__init__.py", line 480, in wait_for_physdevs
        physdevs = extract_physdevs(netcfg)
      File "/usr/lib/python3/dist-packages/cloudinit/net/__init__.py", line 470, in extract_physdevs
        version = netcfg.get('version')
    AttributeError: 'str' object has no attribute 'get'

Its not clear to me how `_network_config` got a string in it.  Your change suggests that it got sources.UNSET, but i'm not sure how that would have happened either.  I'd like to understand how it got set to that.

If we want to use 'UNSET' instead of None to mean unset (it appears some other datasources do that) then thats fine, but it should then be *initialized* to sources.UNSET in the __init__.

**Comment by raharper** [comment]
> the stack trace in question is this:
> 
> ```
> 2019-12-20 09:57:05,749 - util.py[WARNING]: failed stage init
> 2019-12-20 09:57:05,750 - util.py[DEBUG]: failed stage init
> Traceback (most recent call last):
>   File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 653, in status_wrapper
>     ret = functor(name, args)
>   File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 362, in main_init
>     init.apply_network_config(bring_up=bool(mode != sources.DSMODE_LOCAL))
>   File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 701, in apply_network_config
>     net.wait_for_physdevs(netcfg)
>   File "/usr/lib/python3/dist-packages/cloudinit/net/__init__.py", line 480, in wait_for_physdevs
>     physdevs = extract_physdevs(netcfg)
>   File "/usr/lib/python3/dist-packages/cloudinit/net/__init__.py", line 470, in extract_physdevs
>     version = netcfg.get('version')
> AttributeError: 'str' object has no attribute 'get'
> ```
> 
> Its not clear to me how `_network_config` got a string in it. Your change suggests that it got sources.UNSET, but i'm not sure how that would have happened either. I'd like to understand how it got set to that.
> 
> If we want to use 'UNSET' instead of None to mean unset (it appears some other datasources do that) then thats fine, but it should then be _initialized_ to sources.UNSET in the **init**.

Looking closer at the logs; the logic flow is like this:

init-local writes out a simple network config:
local calls _write_to_cache() which renders obj.pkl
local exits
init runs
init calls _restore_from_checked_cache()
which then verifies that the current instance_id on disk matches the datasource instanceid, so we do restore from cache; with the network_config from local.

_however_

When stages runs update_metadata(), this  invokes sources.clear_cached_attrs under the "network" scope which will reset the cached value (in our case, valid network_config) to UNSET.

```
for scope, matched_events in supported_events.items():                 
            LOG.debug(                                                         
                "Update datasource metadata and %s config due to events: %s",  
                scope, ', '.join(matched_events))                              
            # Each datasource has a cached config property which needs clearing
            # Once cleared that config property will be regenerated from       
            # current metadata.                                                
            self.clear_cached_attrs((('_%s_config' % scope, UNSET),)) 
```

So, self._network_config _IS_ being reset to UNSET.  With that in mind, the change does indeed make more sense.

The change will get things going again, but I don't want to have to do this sort of digging again; so we need to step-back a bit and find a way to log better messages.  I'd like to see a more common pattern for Datasources to use around network_config, clear indication of whether we've restored from cachek, the current values of datasource.network_config / datasource._network_config.

**Comment by smoser** [comment]
> Looking closer at the logs; the logic flow is like this:
> 
> init-local writes out a simple network config:
> local calls _write_to_cache() which renders obj.pkl
> local exits
> init runs
> init calls _restore_from_checked_cache()
> which then verifies that the current instance_id on disk matches the datasource instanceid, so we do restore from cache; with the network_config from local.
> 
> _however_
> 
> When stages runs update_metadata(), this invokes sources.clear_cached_attrs under the "network" scope which will reset the cached value (in our case, valid network_config) to UNSET.

This seems wrong.  there *was* a value found, it shouldn't get cleared.  This is a 'local' datasource, so the networking information it declared should be kept and re-used. 
> 
> ```
> for scope, matched_events in supported_events.items():                 
>             LOG.debug(                                                         
>                 "Update datasource metadata and %s config due to events: %s",  
>                 scope, ', '.join(matched_events))                              
>             # Each datasource has a cached config property which needs clearing
>             # Once cleared that config property will be regenerated from       
>             # current metadata.                                                
>             self.clear_cached_attrs((('_%s_config' % scope, UNSET),)) 
> ```
> 
> So, self._network_config _IS_ being reset to UNSET. With that in mind, the change does indeed make more sense.

So then at least 2 things

 1. Set the initial value to unset (rather than None)
 2. datasource needs to properly set the value of _network_config, so that it doesn't get re-set by the clear_cached_attrs... or there is a bug there.

**Comment by karibou** [comment]
Hello,

In Scaleway's context, the clearing of the previous _network_config is what we want as the IPv6 address may change between reboots, hence the addition of **EventType.BOOT** to **update_events**.

With that in mind, @smoser item 1 stands, and the PR should be changed to :
`if self._network_config != sources.UNSET:
    return self._network_config`

Sounds correct to you ?

**Comment by smoser** [comment]
@karibou I had forgotton/not realized about eventtype.boot.

Your suggestion does seem correct, but you'll still need to
a.) set the value in __init__ to be sources.UNSET rather than:

    self._network_config = None

b.) Maybe (i'm not sure) have to account for the fact old restored-from-cache data might have None cached.   I'd *rather* forget about the fact that it ever could have been None, so please test and lets see if we can do that.  You could even LOG.Warning if it *was* None.

**Comment by karibou** [comment]
Hi @smoser,
I just updated the branch with your suggested changes.

...Louis

**Comment by karibou** [comment]
Hello,
For some reason my branch seems out of date even though a force push after rebasing master doesn't change its state.
@smoser I have run the requested tests and in both cases no WARNING appears in the logs.
I also updated the comment in the test case as requested.

Let me know if anything else is needed.
Kind regards,
...Louis

**Comment by karibou** [comment]
Hi @smoser ,
FYI, the impact of this issue seems more important on Ubuntu Focal. It makes the rest of cloud-init execution fail. As expected, this MR's fix does correct the situation.

...Louis

**Comment by smoser** [comment]
this looks good. Thanks for sticking with it, Louis.

fwiw, I hit the "update merge" button that github web ui provided. I think if you had rebased, you had not udpated your upstream (origin) repo.

Given the 'origin' remote as github.com/canonical/cloud-init and the 'karibou' remote as github.com/karibou/cloud-init, you do something like this:

     $ git fetch origin
     $ git fetch karibou
     $ git checkout karibou/fix_scaleway_datasource_backtrace
     $ git rebase -i origin/master
     # this went automatically for me
     $ git push --force karibou HEAD:fix_scaleway_datasource_backtrace

**Pull request merged**